### ISSUE TITLE
Fog is weird sometimes

### ISSUE THREAD

**Issue by numberZero** [minetest/minetest#5128]
![blue fog](https://i.imgur.com/9JVkPlB.png)
![more blue fog](https://i.imgur.com/yrfg8BF.png)
Bright blue intermixed with deep black...
Shaders: enabled
Filters: mipmap, anisotropic (unsure)
Tone mapping: enabled

**Comment by paramat** [comment]
What GPU? I have Intel integrated and mip-map causes visual bugs for me.

**Comment by numberZero** [comment]
NVIDIA, not Intel. `lspci` tells it is:
`01:00.0 VGA compatible controller: NVIDIA Corporation G98 [GeForce 8400 GS Rev. 2] (rev a1)`

**Comment by nerzhul** [comment]
@juhdanad if i remember you worked on this part ?

**Comment by juhdanad** [comment]
@numberZero Can you still see the bug if you turn off shaders?

**Comment by numberZero** [comment]
@juhdanad no

**Comment by numberZero** [comment]
@juhdanad This seems to fix the problem:
```diff
diff --git a/client/shaders/nodes_shader/opengl_fragment.glsl b/client/shaders/nodes_shader/opengl_fragment.glsl
index 914f9f87..7c5b9b61 100644
--- a/client/shaders/nodes_shader/opengl_fragment.glsl
+++ b/client/shaders/nodes_shader/opengl_fragment.glsl
@@ -37,7 +37,7 @@ const float fogShadingParameter = 1 / ( 1 - fogStart);
 
 vec3 uncharted2Tonemap(vec3 x)
 {
-       return ((x * (0.22 * x + 0.03) + 0.002) / (x * (0.22 * x + 0.3) + 0.06)) - 0.03334;
+       return ((x * (0.22 * x + 0.03) + 0.002) / (x * (0.22 * x + 0.3) + 0.06)) - 0.03333;
 }
 
 vec4 applyToneMapping(vec4 color)
```

**Comment by juhdanad** [comment]
Then could you make a pull request? I did not modify that part of the code, and do not know why is it wrong.

**Comment by paramat** [comment]
@numberZero why does that fix it, maybe because it avoids making ther eturn value negative?

**Comment by HybridDog** [comment]
I assume that the problem results from the little negative number:
```lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
> function f(x) return ((x * (0.22 * x + 0.03) + 0.002) / (x * (0.22 * x + 0.3) + 0.06)) - 0.03334  end
> = f(0)
-6.6666666666687e-06
> function f(x) return ((x * (0.22 * x + 0.03) + 0.002) / (x * (0.22 * x + 0.3) + 0.06)) - 0.03333  end
> return  f(0)
3.3333333333344e-06
```
The problem likely only happens with enabled tonemapping.

**Comment by paramat** [comment]
Yes that was my suspicion. I might make a PR for this.

**Comment by paramat** [comment]
Already fixed in #5146 , i should have seen the link above.
